kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,268 @@
+    /// should be used instead.
+    fn read_range(
+        &self,
+        range: &AddressRange,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Read data from blocks starting at `block`, and into the buffer.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// On success, triggers `Client::read_complete` once.
+    fn read(
+        &self,
+        region: &Region<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Write data from a buffer to flash.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
+    /// The user of this function SHOULD ensure that a block is erased
+    /// before calling `write` (see `erase`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it SHOULD NOT be written again until it's erased as part of an erase block.
+    /// Multiple writes to the same block are possible in this trait,
+    /// but the result is dependent on the nature of the underlying hardware,
+    /// and this interface does not provide a way to discover that.
+    ///
+    /// **Note** about raw flash devices: writes can turn bits from `1` to `0`.
+    /// To change a bit from `0` to `1`, a region must be erased.
+    ///
+    /// **Note**: some raw flash hardware only allows a limited number of writes before
+    /// an erase. If that is the case, the implementation MUST return an error
+    /// `ErrorCode::NOMEM` when this happens, even if the hardware silently
+    /// ignores the write.
+    ///
+    /// On success, triggers `Client::write_complete` once.
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    fn write(
+        &self,
+        region: &Region<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Makes a region ready for writing.
+    ///
+    /// This corresponds roughly to the erase operation on raw flash devices,
+    /// but may also do nothing on devices where erasure is not necessary.
+    ///
+    /// Calling `erase` may modify bytes in the selected `region` in any way
+    /// (typically, on raw flash devices, all bits will be set,
+    /// i.e. each byte turns to 0xFF).
+    ///
+    /// If `region` exceeds the size of the device, returns `ErrorCode::INVAL`.
+    /// On success, triggers `Client::erase_complete` once.
+    fn erase(&self, region: &Region<E>) -> Result<(), ErrorCode>;
+
+    /// Returns the size of the device in bytes.
+    fn get_size(&self) -> Result<u64, ErrorCode>;
```

Comment:
How could this fail?

Reply:
I think this could be reasonably expected to ask the device to identify itself, and fail if the size can't be determined.

On the other hand, perhaps it could be made infallible. The other methods are already supposed to fail if bounds are exceeded, and it doesn't make a lot of sense to not know how big a device is from the init.

I could use some feedback on this.

Reply:
Well if this leads to a device query then it would have to be asynchronous, or the device query would have to happen in advance which complicates the implementation.

But in general, are there hardware designs which wouldn't know the size of their block storage? It seems like it's a reasonable assumption for an embedded device to make that it knows the size of its storage.

Reply:
I removed the possibility to have an error completely. The data manipulation needs to know the device size anyway to be able to error out early anyway.